multiversx/mx-chain-go
# Feat/user tx pool

## Reasoning behind the pull request
- When a relayed tx is received, /transaction/pool API by sender does not know about the user tx that may be from that sender
  
## Proposed changes
- Added new sharded cacher that holds the user transactions with 2 keys, the sender and the parent tx hash. It hooks to the main transaction cacher, so it cleans the transactions once the parent relayed is done. With this feature, /transaction/pool API, when used with by-sender parameter, is able to return the inner transaction sent by the specified sender as well.

## Testing procedure
- standard system test
- send a relayed tx with a user tx that should stay in the pool(eg the user tx has a higher nonce). Call /transaction/pool?by-sender= with the user tx sender. For a short period of time, this should return the inner tx with its parent hash on hash field. (&fields=data,sender,receiver,nonce can be used for more details about it). **Note:** this scenario with short period of time only applies when the user tx sender is in a different shard than relayer.
- also to be checked the scenarios:
   - relayer in shard 0, user tx sender in shard 1, user tx receiver in shard 0
   - all 3 participants in the same shard, nonce ok on inner tx, but higher nonce on relayed tx, then search tx pool by user tx sender should return the tx and have it for a longer period

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?
